sklearn/linear_model/_linear_loss.py

```diff
@@ -0,0 +1,411 @@
+        """Helper function to get coefficients, intercept and raw_prediction.
+
+        Parameters
+        ----------
+        coef : ndarray of shape (n_dof,), (n_classes, n_dof) or (n_classes * n_dof,)
+            Coefficients of a linear model.
+            If shape (n_classes * n_dof,), the classes of one feature are contiguous,
+            i.e. one reconstructs the 2d-array via
+            coef.reshape((n_classes, -1), order="F").
+        X : {array-like, sparse matrix} of shape (n_samples, n_features)
+            Training data.
+
+        Returns
+        -------
+        weights : ndarray of shape (n_features,) or (n_classes, n_features)
+            Coefficients without intercept term.
+        intercept : float or ndarray of shape (n_classes,)
+            Intercept terms.
+        raw_prediction : ndarray of shape (n_samples,) or \
+            (n_samples, n_classes)
+        """
+        if not self._loss.is_multiclass:
+            if self.fit_intercept:
+                intercept = coef[-1]
+                weights = coef[:-1]
+            else:
+                intercept = 0.0
+                weights = coef
+            raw_prediction = X @ weights + intercept
+        else:
+            # reshape to (n_classes, n_dof)
+            if coef.ndim == 1:
+                weights = coef.reshape((self._loss.n_classes, -1), order="F")
+            else:
+                weights = coef
+            if self.fit_intercept:
+                intercept = weights[:, -1]
+                weights = weights[:, :-1]
+            else:
+                intercept = 0.0
+            raw_prediction = X @ weights.T + intercept  # ndarray, likely C-contiguous
+
+        return weights, intercept, raw_prediction
+
+    def loss(self, coef, X, y, sample_weight=None, l2_reg_strength=0.0, n_threads=1):
+        """Compute the loss as sum over point-wise losses.
+
+        Parameters
+        ----------
+        coef : ndarray of shape (n_dof,), (n_classes, n_dof) or (n_classes * n_dof,)
+            Coefficients of a linear model.
+            If shape (n_classes * n_dof,), the classes of one feature are contiguous,
+            i.e. one reconstructs the 2d-array via
+            coef.reshape((n_classes, -1), order="F").
+        X : {array-like, sparse matrix} of shape (n_samples, n_features)
+            Training data.
+        y : contiguous array of shape (n_samples,)
+            Observed, true target values.
+        sample_weight : None or contiguous array of shape (n_samples,), default=None
+            Sample weights.
+        l2_reg_strength : float, default=0.0
+            L2 regularization strength
+        n_threads : int, default=1
+            Number of OpenMP threads to use.
+
+        Returns
+        -------
+        loss : float
+            Sum of losses per sample plus penalty.
+        """
+        weights, intercept, raw_prediction = self._w_intercept_raw(coef, X)
+
+        loss = self._loss.loss(
```

Comment:
The choice of names is confusing:
The class is a loss function object called `LinearModelLoss`, with a method `loss`, an init parameter `loss`, and an attribute `self._loss`. The init parameter `loss` contains another loss function object (e.g. `HalfBinomialLoss`). It is stored in `self._loss`. The method `loss` calls the method `self._loss.loss` to get a result called `loss`, which is summed over samples into a new variable `loss`. Note that the method `loss` does not directly returns `loss`, but adds the L2 normalization beforehand.

To improve readability:
- In this method, the variable names could be improved without changing the API: `loss_per_sample`, `sum_loss`, `loss_value`, ...
- Maybe the private attribute could be rename into `self._loss_function`, or `self._sub_loss`.
- Is there a conceptual difference between objects like `HalfBinomialLoss` and `LinearModelLoss`? The names imply that they have the same API. Can I use `HalfBinomialLoss` directly instead of `LinearModelLoss(loss=HalfBinomialLoss)` ? Can I call `LinearModelLoss(loss=LinearModelLoss(loss=HalfBinomialLoss))`?